### PR TITLE
Docs for Timelock Service - Phase 1

### PR DESCRIPTION
**Pull request by jeremyk-91**
This includes a fair amount of documentation for the external Timelock Service:

* Why External Timelock Servers?
* Installation process
* Server- and Client-side configuration (inclusive of invariants)
* Manual migration procedures

**NOT** included:

* Because they are not a part of PR #1255 which I seek to get merged in the near-term (though they must be done soon):
  - Downgrade / Rollback CLI
  - Automated upgrades
* Because they aren't high priority for internal use:
  - Architecture
  - Guarantees of the Timestamp and Lock Services (which I would like to improve upon)
  - Guarantees of External Timelock Services (again, I'd like to write this, but not now)

*Do not merge* until PR #1255 (which includes the fast forward endpoint and client config) is merged. Post changes to that PR in review, details of the configuration or migration that have changed in code may need to be changed here as well.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1279)

### CODE REVIEW

File: docs/source/services/timelock_service/index.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+.. warning::
+
+   Running external timestamp and lock services is currently an experimental feature.
+   If you are interested in testing it, please contact the AtlasDB team.
+
+.. danger::
+
+   Improperly configuring one's cluster to use external timestamp and lock services can result in **SEVERE DATA
+   CORRUPTION**! Please contact the AtlasDB team if you wish to try this feature.
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/index.rst]: This is different to the warning in the file above. Intentional?

File: docs/source/services/timelock_service/index.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    installation
+    server_configuration
+    client_configuration
+    migration
+
+The AtlasDB Timelock Service is an external implementation of the Timestamp and Lock services. Running an external
+Timelock Service (as opposed to running timestamp and lock services embedded within one's clients) provides several
+benefits:
+
+- Additional reliability; we rely on the Atomix_ open-source distributed coordination library,
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/index.rst]: I think we usually do inline links (I guess you'll know more though :) )

File: docs/source/services/timelock_service/installation.rst [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+Timelock Server Installation
+============================
+
+1. Clone the git repository.
+
+   .. code:: bash
+
+      git clone git@github.com:palantir/atlasdb.git; cd atlasdb
+
+2. Generate a tar file with Gradle:
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/installation.rst]: We should be publishing this somewhere. @jboreiko for more info on how we are going to do this. (We did have it in the codebase but it broke internal publishing, and maybe external publishing)

File: docs/source/services/timelock_service/installation.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  This will place a tar file in the ``build/distributions`` directory of the ``atlasdb-timelock-server`` project. The
+  tar file follows semantic versioning.
+
+3. Copy this tar file to the server on which you want to deploy Timelock Server.
+   (If you wish to run Timelock Server locally, this step may be omitted.)
+
+   .. code:: bash
+
+      scp atlasdb-timelock-server/build/distributions/atlasdb-timelock-server-0.25.0.sls.tgz <DESTINATION_HOST>
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/installation.rst]: We probably don't want to list `scp` commands, etc. We should basically say configure the timelock server by editing the file at `var/conf/timelock.yml` and you can start the timelock server with `service/bin/init.sh start`

File: docs/source/services/timelock_service/installation.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+6. Start the Timelock Server.
+
+   .. code:: bash
+
+      service/bin/init.sh start
+
+   This should output the process ID of the Timelock Server. You can view the logs in the (by default) ``var/log``
+   directory.
+
+7. It may be useful to run a health-check to verify that the Timelock Server is running. To do this, you can curl
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/installation.rst]: There's a healthcheck endpoint on the admin port which should be queried instead

File: docs/source/services/timelock_service/migration.rst [hunk 6]
```diff
@@ -0,0 +1,110 @@
+.. _timelock-migration:
+
+Migration to External Timelock Services
+=======================================
+
+.. danger::
+
+   Improperly configuring one's cluster to use external timestamp and lock services can result in **SEVERE DATA
+   CORRUPTION**! Please contact the AtlasDB team before attempting a migration.
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/migration.rst]: This is also different to the ones above.

File: docs/source/services/timelock_service/server_configuration.rst [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+Further Configuration Parameters
+--------------------------------
+
+The Timelock Server is implemented as a Dropwizard application, and may thus be suitably configured_ with a ``server``
+block. This may be useful if, for example, one needs to change the application and/or admin ports for the Timelock
+Server.
+
+.. _Atomix: http://atomix.io/
+.. _configured: http://www.dropwizard.io/0.9.2/docs/manual/configuration.html
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/server_configuration.rst]: 0.9.3?

File: docs/source/services/timelock_service/server_configuration.rst [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+       Atomix uses to store the timestamps and leader state.
+
+       .. warning::
+          If you use the ``MEMORY`` storage level, system failures may result in irrecoverable data loss. This setting
+          is thus highly discouraged outside of test purposes.
+
+   * - storageDirectory
+     - A path corresponding to the location in which Atomix will store its state machine (default: ``var/data/atomix``).
+
+   * - sslConfiguration
```
Comment (SerialVelocity) [inline docs/source/services/timelock_service/server_configuration.rst]: `security`

File: docs/source/services/timelock_service/index.rst (hunk 1 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/index.rst]: Oops, the ending bit of the previous one should say "if you're configuring the client" - good catch.

You can set up the timelock server and play with it without a risk of severe data corruption - the risky bits start when you actually point a cluster at it, or you run the migration commands :)

File: docs/source/services/timelock_service/installation.rst (hunk 4 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/installation.rst]: Sure, will remove this one and the `vi`

File: docs/source/services/timelock_service/installation.rst (hunk 5 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/installation.rst]: OK, will change this

File: docs/source/services/timelock_service/migration.rst (hunk 6 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: This one's intentional (the last few words of the warning are aimed specifically at the feature being documented)

File: docs/source/services/timelock_service/server_configuration.rst (hunk 8 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/server_configuration.rst]: yep, good catch

File: docs/source/services/timelock_service/server_configuration.rst (hunk 7 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/server_configuration.rst]: Right, yep it should be 0.9.3

File: docs/source/services/index.rst [hunk 9]
```diff
@@ -3,12 +3,16 @@ Services
 ========
 
 AtlasDB relies on three distinct services. They are the Key Value Service, the Lock Service, and the Timestamp Service.
-Each of the services is described in detail below.  The intention of this documentation is to provide a high level overview of the architecture of
-each of the services as well as the guarentees they uphold.
+Each of the services is described in detail below.  The intention of this documentation is to provide a high level
+overview of the architecture of each of the services as well as the guarantees they uphold.
+
+In the past, AtlasDB would typically run an embedded timestamp and lock service. We are currently developing an
```
Comment (harrybiddle) [inline docs/source/services/index.rst]: It's not the past yet! I think this would be better as "AtlasDB typically runs ... however, we are currently developing an external time and lock server. Please consult <direct link to docs>"

File: docs/source/services/timelock_service/client_configuration.rst [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+   CORRUPTION**! Please contact the AtlasDB team before you configure your clients to use this.
+
+You will need to update your AtlasDB configuration in order to have said clients request timestamps and locks from
+external Timelock Servers as opposed to their embedded services. This is an extension of the leader block configuration
+options discussed at :ref:`leader-config`.
+
+Instead of configuring a ``leader`` block, or both a ``timestamp`` and ``lock`` block, one may instead specify a
+single ``timelock`` block.
+
+The Timelock Block
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/client_configuration.rst]: Hey, that rhymes!

File: docs/source/services/timelock_service/index.rst [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    client_configuration
+    migration
+
+The AtlasDB Timelock Service is an external implementation of the Timestamp and Lock services. Running an external
+Timelock Service (as opposed to running timestamp and lock services embedded within one's clients) provides several
+benefits:
+
+- Additional reliability; we rely on the `Atomix <http://atomix.io>`__ open-source distributed coordination library,
+  which has been Jepsen tested for fault tolerance.
+- To some extent, individual services can now scale independently of AtlasDB, depending on one's requirements.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/index.rst]: What do you mean by this? That you can add more clients without adding automatically more AtlasDB services? If so, how is this a benefit?

File: docs/source/services/timelock_service/index.rst [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+- Resource contention between your AtlasDB clients and embedded timestamp and lock services can be better managed.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/index.rst]: How can they be better managed? Because they run in separate JVMs and possibly on separate boxes?

File: docs/source/services/timelock_service/index.rst [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+- Timestamp and Lock endpoints are now hidden from users, making abuse more difficult.
+
+Also, Timelock Services can be run in clustered mode, allowing for high availability (as long as a majority quorum
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/index.rst]: I'd just add this to the bullet points above

File: docs/source/services/timelock_service/migration.rst [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+Why Migration?
+--------------
+
+AtlasDB assumes that timestamps returned by the timestamp service are monotonically increasing. In order to preserve
+this guarantee when moving from an embedded timestamp service to an external timestamp service, we need to ensure
+that any timestamps the external timestamp service issues must be larger than any timestamps that the embedded
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: This is a bit of tongue twister! How about "we need to ensure that timestamps issued by the external timestamp service are larger than those issued by the internal one"?

File: docs/source/services/timelock_service/migration.rst [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+service issues. Otherwise, this can lead to serious data corruption.
+
+The migration process must be run offline (that is, with no AtlasDB clients running during migration) and basically
+consists of the following steps:
+
+1. Set up the external timestamp service.
+2. Obtain a timestamp TS that is at least as large as the largest timestamp issued.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: What is a timestamp TS? How do I find the largest timestamp issued?

Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: Ah, I get it! "TS" is referred to below, and you also dig into how to get the largest timestamp issued. We could format like "Obtain a timestamp `TS` that is at least..." to make it clearer that it's a variable you're using.

File: docs/source/services/timelock_service/migration.rst [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+Configure your clients to use the Timelock Server following the instructions in :ref:`timelock-client-configuration`.
+You may then restart your clients; they should now communicate with the Timelock Server when requesting timestamps
+and locks. This completes the migration process.
+
+Automated Migration
+-------------------
+
+The AtlasDB team is currently working on an automated migration process, such that the steps above are run when one
+initiates a ``TransactionManager`` with a timelock configuration for the first time.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: I see that you're intentionally not including details about rollback yet; but it might be worth mentioning it, in the same way that you mentioned automated migration, so that people know it is on our plate.

File: docs/source/services/timelock_service/migration.rst [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+   Make sure that N has been entered correctly.
+
+    - If N entered is too large, there will generally not be adverse consequences (apart from risks of ``long``
+      overflow), but...
+    - If N entered is too small and AtlasDB clients are restarted without rectifying this, this can result in
+      **SEVERE DATA CORRUPTION** because we lose the guarantee that timestamps are monotonically increasing.
+      As fast-forward is idempotent (it sets the timestamp to be the maximum of its current value and the
+      ``newMinimum``), *if clients have not been started again* this can be fixed by doing another fast-forward.
+
+Step 4: Invalidating the Old Atlas Timestamp
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: AtlasDB 💃

File: docs/source/services/timelock_service/migration.rst [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+not typically exposed to AtlasDB clients. One can use it to advance the timestamp on the Timelock Server to N, as
+follows:
+
+   .. code:: bash
+
+      curl -XPOST localhost:8080/test/timestamp-admin/fast-forward?newMinimum=N
+
+.. danger::
+
+   Make sure that N has been entered correctly.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: What happens if I get N wrong? Can I go back again, or is it a one-way street?

File: docs/source/services/timelock_service/migration.rst [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+      **SEVERE DATA CORRUPTION** because we lose the guarantee that timestamps are monotonically increasing.
+      As fast-forward is idempotent (it sets the timestamp to be the maximum of its current value and the
+      ``newMinimum``), *if clients have not been started again* this can be fixed by doing another fast-forward.
+
+Step 4: Invalidating the Old Atlas Timestamp
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The steps for invalidating the old Atlas timestamp will vary, depending on your choice of underlying key value store.
+
+- Dropping the table, generally speaking, will *not* work (on the next startup of an embedded Timestamp Service,
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: I think it would be better to start with what _does_ work, clearly for each DB. And then dispell the myths of what won't afterwards. I basically think when people read this, they'll look for their KVS, look for your advice, and then perhaps wonder "oh, why can't I just drop the table?", in that order.

File: docs/source/services/timelock_service/migration.rst [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+  AtlasDB will believe it is starting up the Timestamp Service for the first time, and thus start again from 1).
+- Setting the value to ``Long.MAX_VALUE`` or ``Long.MIN_VALUE`` will not work (Java Longs do not throw on arithmetic
+  overflow, and although ordinarily the first timestamp AtlasDB issues is 1 we do not throw on negative numbers).
+- If using Postgres or Oracle, it suffices to rename the relevant column in the timestamp table (use ``ALTER TABLE``).
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: Are we able to give the whole command?

File: docs/source/services/timelock_service/migration.rst [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+- If using Cassandra, one method of invalidating the table is to overwrite the timestamp bound record with the
+  empty byte array (consider using ``cqlsh`` to do this):
+
+     .. code:: bash
+
+        INSERT INTO atlasdb."_timestamp" (key, column1, column2, value) VALUES (0x7473, 0x7473, -1, 0x);
+
+  You may want to save the old value of the timestamp, in the event that rollback is desired.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: I think we should just tell people that they should always save the old value, and give the command to get this.

File: docs/source/services/timelock_service/migration.rst [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+
+Please contact the AtlasDB team for assistance if you are uncertain about this step or otherwise run into difficulties.
+
+Steps 5 and 6: Client Configuration and Restart
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Configure your clients to use the Timelock Server following the instructions in :ref:`timelock-client-configuration`.
+You may then restart your clients; they should now communicate with the Timelock Server when requesting timestamps
+and locks. This completes the migration process.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/migration.rst]: Is there any verification people can do, so that they know everything has gone OK?

File: docs/source/services/timelock_service/server_configuration.rst [hunk 23]
```diff
@@ -0,0 +1,101 @@
+.. _timelock-server-configuration:
+
+Timelock Server Configuration
+=============================
+
+The Timelock Server configuration file is written in YAML and is located at ``var/conf/timelock.yml``.
+It has three main blocks: the ``clients`` block, ``cluster`` block and ``atomix`` block. We will discuss how each of
+these may be configured in turn, as well as additional configuration parameters.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/server_configuration.rst]: Nice intro! Really clear

File: docs/source/services/timelock_service/server_configuration.rst [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+   * - servers
+     - A list of strings following the form ``protocol://hostname:port`` identifying the hosts in this Timelock
+       Service cluster. Note that this list must include the ``localServer``.
+
+Atomix
+------
+
+The Timelock Servers use the `Atomix <http://atomix.io>`__ library, and allow for some configuration as to
+Atomix-related communication and persistence. Note that unlike the ``clients`` and ``cluster`` blocks, this block is
+optional.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/server_configuration.rst]: I think it would be worth having example YAML here too. I really appreciate just an example I can look at, and just copy/paste into my own `timelock.yml`. Presumably it looks something like this?

```
atomix:
    storageLevel: DISK
    storageDirectory: var/data/atomix
...etc
```

File: docs/source/services/timelock_service/server_configuration.rst [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+   * - storageLevel
+     - One of ``DISK``, ``MEMORY`` or ``MAPPED`` (default: ``DISK``). These correspond to the level of persistence
+       Atomix uses to store the timestamps and leader state.
+
+       .. warning::
+          If you use the ``MEMORY`` storage level, system failures may result in irrecoverable data loss. This setting
+          is thus highly discouraged outside of test purposes.
+
+   * - storageDirectory
+     - A path corresponding to the location in which Atomix will store its state machine (default: ``var/data/atomix``).
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/server_configuration.rst]: Is this relative to the root directory that the timelock server was unpacked at? Does an absolute path work?

File: docs/source/services/timelock_service/server_configuration.rst [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+   * - security
+     - Security settings for communication between Atomix nodes, following the
+       `palantir/http-remoting <https://github.com/palantir/http-remoting/blob/develop/ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslConfiguration.java>`__
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/server_configuration.rst]: Is there any example YAML we can link to?

File: docs/source/services/timelock_service/client_configuration.rst (hunk 10 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/client_configuration.rst]: Not to knock myself too hard, but I didn't put too much stock in having the phrasing rock :)

File: docs/source/services/timelock_service/index.rst (hunk 11 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/index.rst]: Yep, it means that you can have different numbers of clients and AtlasDB servers. This could be useful if you have large numbers of clients, since you have a fair bit of message complexity with all of the heartbeats going on if you maintained a distributed leader.

File: docs/source/services/timelock_service/index.rst (hunk 13 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/index.rst]: Added.

File: docs/source/services/timelock_service/migration.rst (hunk 14 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Thanks - have changed it to that (I use "embedded" rather than "internal" to be consistent, though).

File: docs/source/services/timelock_service/migration.rst (hunk 15 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Good idea - thanks!

File: docs/source/services/timelock_service/migration.rst (hunk 18 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Okay. Basically if N is too large that's okay (apart from `Long.MAX_VALUE` stuff! - though there's a way around that).
If N is too small you need to fast forward again (if clients haven't been started again; if they have, you could do it but you'll already have some data corruption). 

I've mentioned this in the block that follows, but any suggestions on how / if I could make this clearer?

File: docs/source/services/timelock_service/migration.rst (hunk 17 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Fixed this - also fixed the Step 2 one.

File: docs/source/services/timelock_service/migration.rst (hunk 19 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Reordered to put the positive cases first.

File: docs/source/services/timelock_service/migration.rst (hunk 18 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: (Also on second thought I've updated everything to say TS instead of N)

File: docs/source/services/timelock_service/migration.rst (hunk 20 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Yep - I've added the full command

File: docs/source/services/timelock_service/migration.rst (hunk 22 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: I've added verification steps at the end of Step 3 and Step 4; hopefully that's enough. I'm not sure what to say for these steps other than "check that your app still works!"

File: docs/source/services/timelock_service/migration.rst (hunk 16 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: I've added a brief section at the end.

File: docs/source/services/timelock_service/server_configuration.rst (hunk 24 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/server_configuration.rst]: Good call - I've added it too. Not sure why I left it out initially!

File: docs/source/services/timelock_service/server_configuration.rst (hunk 25 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/server_configuration.rst]: Yep, this is relative to the root directory. I'll add a bit of clarifying information

File: docs/source/services/timelock_service/server_configuration.rst (hunk 26 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/server_configuration.rst]: Hmm, I didn't find one in the SSL config repository, unfortunately - the closest I found was their [SerializationTest](https://github.com/palantir/http-remoting/blob/e8886ff942c3f168519a695a776d899df9edf0ad/ssl-config/src/test/java/com/palantir/remoting1/config/ssl/SerializationTests.java), but that used JSON, not YAML unfortunately.

Comment (jeremyk-91) [inline docs/source/services/timelock_service/server_configuration.rst]: (I refrained from providing examples, in case this changed under us - do you think it would be worthwhile including one?)

File: docs/source/services/timelock_service/index.rst (hunk 1 above)
Comment (rhero) [inline docs/source/services/timelock_service/index.rst]: Can we combine these two into one "Danger" warning? I assume we're planning on removing these warning and danger notes anyway once this is the recommended default with proper safeguards.

File: docs/source/services/timelock_service/installation.rst [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+   (If you wish to run Timelock Server locally, this step may be omitted.)
+
+4. On the server which you intend to deploy Timelock Server, untar the archive.
+
+   .. code:: bash
+
+      tar -zxvf atlasdb-timelock-server-0.25.0.sls.tgz; cd atlasdb-timelock-server-0.25.0
+
+5. Configure the Timelock Server - see :ref:`timelock-server-configuration` for a guide on this. The configuration file
+   is at ``/var/conf/timelock.yml``.
```
Comment (rhero) [inline docs/source/services/timelock_service/installation.rst]: should this be a relative path?

File: docs/source/services/timelock_service/installation.rst [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+   directory.
+
+7. It may be useful to run a health-check to verify that the Timelock Server is running. To do this, you can curl
+   the server's ``healthcheck`` endpoint on its admin port (default: 8081).
+
+   .. code:: bash
+
+      curl localhost:8081/healthcheck
+
+   The output should indicate that the Timelock Server is healthy.
```
Comment (rhero) [inline docs/source/services/timelock_service/installation.rst]: can you add what the healthcheck returns if it's short?

File: docs/source/services/timelock_service/server_configuration.rst [hunk 29]
```diff
@@ -0,0 +1,108 @@
+.. _timelock-server-configuration:
```
Comment (rhero) [inline docs/source/services/timelock_service/server_configuration.rst]: I feel like the client and server configuration pages should go in the larger "Configuration" section of our docs. If I'm trying to configure the timelock server, the first place I'll check is probably Configuration -> Timelock -> client|server. Going to Services -> External Timelock Service -> *configuration feels less intuitive.

File: docs/source/services/timelock_service/server_configuration.rst [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+Timelock Server Configuration
+=============================
+
+The Timelock Server configuration file is written in YAML and is located at ``var/conf/timelock.yml``.
+It has three main blocks: the ``clients`` block, ``cluster`` block and ``atomix`` block. We will discuss how each of
+these may be configured in turn, as well as additional configuration parameters.
+
+.. contents::
+   :local:
+
```
Comment (rhero) [inline docs/source/services/timelock_service/server_configuration.rst]: I would like to see an example of a complete, valid server config. Is it possible to embed an actual file from our codebase?

File: docs/source/services/timelock_service/migration.rst [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+
+Why Migration?
+--------------
+
+AtlasDB assumes that timestamps returned by the timestamp service are monotonically increasing. In order to preserve
+this guarantee when moving from an embedded timestamp service to an external timestamp service, we need to ensure
+that timestamps issued by the external timestamp service are larger than those issued by the embedded one.
+Otherwise, this can lead to serious data corruption.
+
+The migration process must be run offline (that is, with no AtlasDB clients running during migration) and basically
```
Comment (rhero) [inline docs/source/services/timelock_service/migration.rst]: We should be explicit in the steps below that "Step 0: Shutdown your AtlasDB clients". We don't want people to forget that, grab a timestamp and then end up breaking the timestamp server guarantees when they restart their client.

File: docs/source/services/timelock_service/migration.rst [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+2. Obtain a timestamp ``TS`` that is at least as large as the largest timestamp issued.
+3. Fast-forward the external timestamp service to at least ``TS``.
+4. Invalidate the old timestamp service, preventing AtlasDB clients from retrieving timestamps from it.
+5. Configure AtlasDB clients to retrieve timestamps from the new timestamp service.
+6. Restart AtlasDB clients.
+
+Strictly speaking, step 4 is not required; that said, we prefer to do this to avoid the risk of data corruption in the
+event of misconfiguration (that is, if some but not all clients are pointed to the new timestamp service).
+
+Manual Migration
```
Comment (rhero) [inline docs/source/services/timelock_service/migration.rst]: Great manual steps! I didn't actually expect us to document this at all given we're automating it, but I think having the logic here is very clear and we can share exactly how the process works with this document alone.

File: docs/source/services/timelock_service/client_configuration.rst [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+------------------
+
+The ``cluster`` block is used to identify the servers which make up a Timelock Service cluster. An example is as
+follows:
+
+   .. code:: yaml
+
+      timelock:
```
Comment (rhero) [inline docs/source/services/timelock_service/client_configuration.rst]: We should add another example to the [leader configuration page](https://palantir.github.io/atlasdb/html/configuration/leader_config.html#leader-configuration-examples) that includes this as an option (with an "experimental" in parentheses).

File: docs/source/services/timelock_service/client_configuration.rst [hunk 34]
```diff
@@ -0,0 +1,54 @@
...
+
+   * - client
+     - A string which indicates the namespace for which the client will request locks and timestamps. This client
+       must also be on the ``clients`` list of the Timelock Server, as discussed in
+       :ref:`timelock-server-configuration`.
+
+   * - serverListConfig.servers
+     - A list of strings following the form ``protocol://hostname:port`` identifying the hosts in the Timelock Service
+       cluster. At least one server must be specified. AtlasDB assumes that the Timelock Servers which utilise the
+       application connectors referred to by these specifications are in a single cluster, as specified by said server
```
Comment (rhero) [inline docs/source/services/timelock_service/client_configuration.rst]: This sentence is really hard to follow.

File: docs/source/services/timelock_service/server_configuration.rst [hunk 35]
```diff
@@ -0,0 +1,108 @@
...
+The ``cluster`` block is used to identify the servers which make up a Timelock Service cluster. An example is as
+follows:
+
+   .. code:: yaml
+
+      cluster:
+        localServer: http://palantir.com:8700
+        servers:
+          - http://palantir.com:8700
+          - http://palantir.com:8701
```
Comment (rhero) [inline docs/source/services/timelock_service/server_configuration.rst]: Should these be https? Just want to make sure that the examples match our defaults for what we actually use in the field.

File: docs/source/services/timelock_service/server_configuration.rst [hunk 36]
```diff
@@ -0,0 +1,108 @@
...
+     - Description
+
+   * - localServer
+     - A string following the form ``protocol://hostname:port`` which matches the host on which this config exists.
+
+   * - servers
+     - A list of strings following the form ``protocol://hostname:port`` identifying the hosts in this Timelock
+       Service cluster. Note that this list must include the ``localServer``.
+
+Atomix
```
Comment (rhero) [inline docs/source/services/timelock_service/server_configuration.rst]: Can we link to the atomix config docs for "more information"? I think they're [here](http://atomix.io/atomix/docs/configuration/).

File: docs/source/services/timelock_service/index.rst (hunk 1 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/index.rst]: Okay, I've merged the warnings. Once this is the recommended it won't be there anymore, as you say.

File: docs/source/services/timelock_service/installation.rst (hunk 27 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/installation.rst]: Oops, good call - yes it's supposed to be relative. I've changed this

File: docs/source/services/timelock_service/installation.rst (hunk 28 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/installation.rst]: Done - it's a json object with 1 field.

File: docs/source/services/timelock_service/migration.rst (hunk 32 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: I think we spoke offline, but the plan was eventually to rebrand this section as migration architecture or something like that, so I went into full detail here. Thanks!

File: docs/source/services/timelock_service/migration.rst (hunk 31 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Okay, I've added a brief section on this.

File: docs/source/services/timelock_service/client_configuration.rst (hunk 34 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/client_configuration.rst]: Simplified to "AtlasDB assumes that the Timelock Servers being pointed at are backed by the same Atomix cluster."

**Comment by jeremyk-91** [comment]
I will move the configs over and change the server config stuff later, heading back to the P0 now that I have logs to analyze

File: docs/source/services/timelock_service/migration.rst [hunk 37]
```diff
@@ -0,0 +1,152 @@
...
+- If using Cassandra, one method of invalidating the table is to overwrite the timestamp bound record with the
+  empty byte array (consider using ``cqlsh`` to do this). We suggest that you save the old value of the timestamp,
+  in the event that rollback is desired (because our invalidation method in this case is more destructive).
+
+     .. code:: bash
+
+        SELECT * FROM atlasdb."timestamp";
+        <note the value returned by this - call this K>
+        INSERT INTO atlasdb."_timestamp" (key, column1, column2, value) VALEUS (0x7472, 0x7472, -1, K);
+        INSERT INTO atlasdb."_timestamp" (key, column1, column2, value) VALUES (0x7473, 0x7473, -1, 0x);
```
Comment (jeremyk-91) [inline docs/source/services/timelock_service/migration.rst]: Oh dear, this code block needs to be updated.

**Comment by jeremyk-91** [comment]
This should be largely up to date with the state of the world now.

File: docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst [hunk 38]
```diff
@@ -0,0 +1,136 @@
...
+
+   * - paxosDataDir
+     - A path corresponding to the location in which Paxos will store its logs (of accepted promises and learned
+       values) (default: ``var/data/paxos``). The Timelock Server will fail to start if this directory does not
+       exist and cannot be created.
+
+   * - sslConfiguration
+     - Security settings for communication between Timelock Servers, following the
+       `palantir/http-remoting <https://github.com/palantir/http-remoting/blob/develop/ssl-config/src/main/java/com/palantir/remoting1/config/ssl/SslConfiguration.java>`__
+       library (default: no SSL).
```
Comment (rhero) [inline docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst]: Is the no SSL by default going to change? I thought we settled on having it enabled? Or are our internal products defaulting to having this enabled, but AtlasDB itself doesn't enable it?

File: docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst [hunk 39]
```diff
@@ -0,0 +1,136 @@
...
+follows:
+
+   .. code:: yaml
+
+      cluster:
+        localServer: palantir.com:8700
+        servers:
+          - palantir.com:8700
+          - palantir.com:8701
+          - palantir.com:8702
```
Comment (rhero) [inline docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst]: Should these be something like the below instead?

palantir-1.com:8700
palantir-2.com:8700
palantir-3.com:8700

As written it looks like we recommend collocating three servers. Also, is that the default port we settled on?

Might be unrelated, but for the server config, we have this example:

            - "host1:3828"
            - "host2:3828"
            - "host3:3828"

Comment (jeremyk-91) [inline docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst]: I've changed to your suggestion - yes, as you've mentioned there might be the risk that someone reads it as collocating them!

The default port we've settled on isn't 8700; Dropwizard's default is 8080 so I've changed this example to that, though internally we might use something different to avoid clashes.

File: docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst (hunk 38 above)
Comment (jeremyk-91) [inline docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst]: As far as I know, internal products defaulting to having this enabled was to be done via our internal deployability tool.

@hsaraogi probably knows this better than I do, but we can change this later with a PR that changes the defaults if this is the case.
